internal/controller/servicetemplate_controller.go

```diff
@@ -0,0 +1,580 @@
+		if err != nil {
+			return fmt.Errorf("failed to get source status from Bucket %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(bucket.Status.Conditions))
+		copy(conditions, bucket.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	case "OCIRepository":
+		ociRepository := &sourcev1beta2.OCIRepository{}
+		err = r.Get(ctx, key, ociRepository)
+		if err != nil {
+			err = fmt.Errorf("failed to get referred OCIRepository %s: %w", key, err)
+			break
+		}
+		status.SourceStatus, err = r.sourceStatusFromFluxObject(ociRepository)
+		if err != nil {
+			return fmt.Errorf("failed to get source status from OCIRepository %s: %w", key, err)
+		}
+		conditions := make([]metav1.Condition, len(ociRepository.Status.Conditions))
+		copy(conditions, ociRepository.Status.Conditions)
+		status.SourceStatus.Conditions = conditions
+	default:
+		err = fmt.Errorf("unsupported source kind %s", ref.Kind)
+	}
+
+	if err != nil {
+		status.TemplateValidationStatus.Valid = false
+		status.TemplateValidationStatus.ValidationError = err.Error()
+	} else {
+		status.TemplateValidationStatus.Valid = true
+	}
+	template.Status = status
+	return err
+}
+
+func (r *ServiceTemplateReconciler) reconcileRemoteSource(ctx context.Context, template *kcm.ServiceTemplate) error {
+	ref := template.RemoteSourceSpec()
+	if ref == nil {
+		return errors.New("remote source ref is undefined")
+	}
+
+	switch {
+	case ref.Git != nil:
+		return r.reconcileGitRepository(ctx, template, ref)
+	case ref.Bucket != nil:
+		return r.reconcileBucket(ctx, template, ref)
+	case ref.OCI != nil:
+		return r.reconcileOCIRepository(ctx, template, ref)
+	}
+	return errors.New("unknown remote source definition")
+}
+
+func (r *ServiceTemplateReconciler) reconcileGitRepository(
+	ctx context.Context,
+	template *kcm.ServiceTemplate,
+	ref *kcm.RemoteSourceSpec,
+) error {
+	// we'll create a new GitRepository object for every ServiceTemplate. This is needed because we cannot reuse the
+	// same GitRepository objects for multiple ServiceTemplates due to possible difference between Git repository parameters.
+	// Since there would be no labels to identify created GitRepository objects, they won't be reconciled by the flux
+	// controller. The OwnerReference will be set to the ServiceTemplate object, so that the GitRepository object will be
+	// deleted when the ServiceTemplate is deleted.
+	repositoryName, err := GenerateSourceName(client.ObjectKeyFromObject(template))
+	if err != nil {
+		return fmt.Errorf("failed to generate GitRepository name: %w", err)
+	}
+	gitRepository := &sourcev1.GitRepository{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      repositoryName,
+			Namespace: template.Namespace,
+		},
+	}
+	key := client.ObjectKeyFromObject(gitRepository)
+	if err := r.Get(ctx, key, gitRepository); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get GitRepository %s: %w", key, err)
+		}
+	}
```

Comment:
Is it necessary before using `CreateOrUpdate`? I think it's better to run it before `template.Status.Valid` update (line 333) to get the actual `gitRepository` status.